api/src/main/java/org/eclipse/microprofile/rest/client/ext/AsyncInvocationInterceptor.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.eclipse.microprofile.rest.client.ext;
+
+/**
+ * Implementations of this interface can intercept asynchronous method
+ * invocations.  The MP Rest Client implementation runtime will obtain instances
+ * of this interface by invoking the <code>newInterceptor</code> method of all
+ * registered <code>AsyncInvocationInterceptorFactory</code> providers.
+ *
+ * The MP Rest Client implementation runtime will invoke the <code>pre</code>
+ * method on the main thread prior to returning execution back to the calling
+ * method of the client interface.  The runtime will invoke the
+ * <code>post</code> method on the secondary thread before the client request is
+ * sent.  The <code>prepareContext</code> method should always be invoked before
+ * the <code>applyContext</code> method is invoked, but due to the nature of
+ * multithreading, it is possible that <code>post</code> method may be invoked
+ * before the <code>prepareContext</code> method has completed.  Care should be
+ * taken when implementing this interface to avoid race conditions and
+ * deadlocks.
+ *
+ * Note that the order in which instances of the
+ * <code>AsyncInvocationInterceptor</code> are invoked are determined by the
+ * priority of the <code>AsyncInvocationInterceptorFactory</code> provider.
+ */
+public interface AsyncInvocationInterceptor {
```

Comment:
I don't see a lot of use for this interceptor the way its defined right now.  Is there a way it can get access to the underlying resource? or some kind of context object with more information?

it seems like its analogous to `ClientRequestFilter` and `ClientResponseFilter`, but I'm not sure I fully grok it.

Reply:
The primary reason for it is to allow implementors to have some flexibility in how they implement async methods (for example, CXF marshalls the request on the main thread while Jersey returns immediately - and does everything on the new thread) but allow for providers (like OpenTracing) to track things in thread locals.  

So an implementation of the factory interface could have `@Context` annotation injection for things like headers, URIInfo, etc. Then the factory could create the actual interceptors with any pertinent data (like OpenTracing info from the main thread, etc.).  Now, if they need to ensure that thread locals get copied over to the new thread, they can do that with the interceptor.

Reply:
The OpenTracing part makes sense.  However, the part of the vendors seems a bit confusing.  Is that an implementation detail?  To be clear, the part that's confusing is the use of a `Factory` that creates the interceptor, instead of the developer/framework simply registering the interceptor directly.

Reply:
Ah, sorry for the confusion.

The reason for the Factory is because most use cases of this provider type is per-request, but the standard lifecycle for providers are singleton (per-app).  By using a factory pattern implementors of the factory/interceptors can use either approach.  For example, the OpenTracing use case really needs to be per-request - something like:

```
class MyFactory implements AsyncInvocationInterceptorFactory {
    public AsyncInvocationInterceptor newInterceptor() {
        return new OpenTracingAsyncInterceptor();
    }
}

class OpenTracingAsyncInterceptor implements AsyncInvocationInterceptor {
    private Span span;
    public void prepareContext() {
        span = SomeOtherClass.getSpanFromThreadLocal();
    }

    public void applyContext() {
        SomeOtherClass.setSpanToThreadLocal(span);
    }
}
```

The `span` in this case must be per-request - otherwise wires could get crossed.  But there could be other use cases (like debug logging) where a per-app singleton would be preferred (to avoid excessive garbage, object pooling, etc.)- i.e.:

```
MyLoggingProvider implements AsyncInvocationInterceptorFactory, AsyncInvocationInterceptor {
    static Logger LOG = Logger.getLogger(MyLoggingProvider.class);
 
    @Context
    UriInfo uriInfo;

    public AsyncInvocationInterceptor newInterceptor() {
        return this;
    }

    public void prepareContext() {
        LOG.finest("invoking request for {0} on original thread {1}", uriInfo.getPath(), Thread.currentThread().getId());
    }

    public void applyContext() {
        LOG.finest("receiving response for {0} on async thread {1}", uriInfo.getPath(), Thread.currentThread().getId());
    }
}
```

Is there a simpler way to enable both of these use cases?